pkg/virt-controller/watch/migration.go

```diff
@@ -1260,6 +1273,56 @@ func (c *MigrationController) updatePDB(old, cur interface{}) {
 	}
 }
 
+type vmimCollection []*virtv1.VirtualMachineInstanceMigration
+
+func (c vmimCollection) Len() int {
+	return len(c)
+}
+
+func (c vmimCollection) Less(i, j int) bool {
+	t1 := &c[i].CreationTimestamp
+	t2 := &c[j].CreationTimestamp
+	return t1.Before(t2)
+}
+
+func (c vmimCollection) Swap(i, j int) {
+	c[i], c[j] = c[j], c[i]
+}
+
+func (c *MigrationController) garbageCollectFinalizedMigrations(vmi *virtv1.VirtualMachineInstance) error {
+
+	var finalizedMigrations []string
+
+	migrations, err := c.listMigrationsMatchingVMI(vmi.Namespace, vmi.Name)
+	if err != nil {
+		return err
+	}
+
+	// Oldest first
+	sort.Sort(vmimCollection(migrations))
+	for _, migration := range migrations {
+		if migration.IsFinal() && migration.DeletionTimestamp == nil {
+			finalizedMigrations = append(finalizedMigrations, migration.Name)
+		}
+	}
+
+	// only keep the oldest 5 finalized migration objects
+	garbageCollectionCount := len(finalizedMigrations) - defaultFinalizedMigrationGarbageCollectionBuffer
+
+	if garbageCollectionCount <= 0 {
+		return nil
+	}
+
+	for i := 0; i < garbageCollectionCount; i++ {
+		err = c.clientset.VirtualMachineInstanceMigration(vmi.Namespace).Delete(finalizedMigrations[i], &v1.DeleteOptions{})
```

Comment:
I think this is another place where we need to use expectations. There will definitely be failing deletes on the next wakeups.

Reply:
It's actually pretty unlikely we'll attempt to delete migrations multiple times. 

The reality is, in production that garbage collection logic is only going to be deleting 1 item usually. For example, think about  a situation where the 6th migraiton object fails, the 1st one is immediately garbage collected... When the 7th fails, the 2nd is garbage collected... and so forth. It's not a massive list except perhaps after an update to this new version.

For the occasional delete error that could technically occur, I know this probably sounds controversial but rather than expectations I think we should suppress errors for already deleted migrations during garbage collection. I'll explain

The "key" we'd use for the expectation would need to be the VMI's key, and not a migration's key. We'd be saying "we expect these garbage collection deletes for this target VMI before we'll allow garbage collection for the VMI to occur again"

The issue here is that we're not guaranteed that we'll ever see the VMI key again for the duration of the controller. This controller works on migration keys, not VMI keys, and we're only guaranteed that we'll see Migration keys again in the event that objects disappear.

The result here is that I can't guarantee that something like a `c.vmiExpectations.DeleteExpectations(vmiKey) ` will ever get called in the migration controller... It probably will in most situations, but i can't guarantee it.

Reply:
With considering deleted object expclititly, what is done now, looks good to me. :+1: